aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -318,35 +277,67 @@ static void isin_sorting(
   }
 }
 
-Tensor where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(condition.device() == self.device() && self.device() == other.device(),
-              "Expected condition, x and y to be on the same device, but condition is on ",
-              condition.device(), " and x and y are on ", self.device(), " and ", other.device(),
-              " respectively");
+namespace {
+Tensor _s_where_dispatch(
+    const Tensor& condition,
+    const Tensor& self,
+    const Tensor& other) {
+  TORCH_CHECK(
+      condition.device() == self.device() && self.device() == other.device(),
+      "Expected condition, x and y to be on the same device, but condition is on ",
+      condition.device(),
+      " and x and y are on ",
+      self.device(),
+      " and ",
+      other.device(),
+      " respectively");
 
   if (condition.scalar_type() == ScalarType::Byte) {
-  TORCH_WARN_ONCE("where received a uint8 condition tensor. This behavior is deprecated and will be removed in a future version of PyTorch. Use a boolean condition instead.");
-} else {
-  TORCH_CHECK(condition.scalar_type() == ScalarType::Bool, "where expected condition to be a boolean tensor, but got a tensor with dtype ", condition.scalar_type());
-}
+    TORCH_WARN_ONCE(
+        "where received a uint8 condition tensor. This behavior is deprecated and will be removed in a future version of PyTorch. Use a boolean condition instead.");
+  } else {
+    TORCH_CHECK(
+        condition.scalar_type() == ScalarType::Bool,
+        "where expected condition to be a boolean tensor, but got a tensor with dtype ",
+        condition.scalar_type());
+  }
 
   c10::MaybeOwned<Tensor> b_condition, b_self, b_other;
-  std::tie(b_condition, b_self, b_other) = expand_outplace(condition, self, other, "where");
+  std::tie(b_condition, b_self, b_other) =
+      expand_outplace(condition, self, other, "where");
   return at::_s_where(*b_condition, *b_self, *b_other);
 }
 
+Tensor get_promoted_tensor(const Tensor& t, ScalarType common_dtype) {
```

Comment:
Add a comment explaining what this function is doing and how it's intended to be used and let's make the name specific to where ("get_promoted_where_tensor"?)

Reply:
This will be helpful. I found it a bit hard to understand the promotion logic here. I would think we always promote both inputs to common_type but that doesn't seems to be what happens here.